File: sdk/python/aip/envelope.py
"""Message envelope helpers"""
import json
from datetime import datetime, timezone
from uuid import uuid4
from typing import Any
from .types import Envelope, MessageType


def create_envelope(
    type: MessageType, from_agent: str, to_agent: str,
    payload: dict[str, Any], *, reply_to: str = "", correlation_id: str = "",
) -> Envelope:
    return Envelope(
        aip="0.1", id=str(uuid4()), type=type,
        from_agent=from_agent, to_agent=to_agent,
        timestamp=datetime.now(timezone.utc).isoformat(),
        payload=payload, reply_to=reply_to, correlation_id=correlation_id,
    )


def canonical_payload(env: Envelope) -> str:
    """Deterministic JSON for signing"""
    return json.dumps({
        "id": env.id, "type": env.type, "from": env.from_agent,
        "to": env.to_agent, "timestamp": env.timestamp, "payload": env.payload,
    }, separators=(",", ":"), sort_keys=False)


def validate_envelope(data: dict[str, Any]) -> bool:
    required = ["aip", "id", "type", "from", "to", "timestamp", "payload"]
    return all(k in data for k in required)

File: sdk/typescript/src/envelope.ts
/**
 * Message envelope — create and validate AIP message envelopes
 */
import { randomUUID } from 'crypto';
import { Envelope, MessageType } from './types';

/** Create a new envelope with defaults filled in */
export function createEnvelope<T>(
  type: MessageType,
  from: string,
  to: string,
  payload: T,
  opts?: { replyTo?: string; correlationId?: string }
): Envelope<T> {
  return {
    aip: '0.1',
    id: randomUUID(),
    type,
    from,
    to,
    timestamp: new Date().toISOString(),
    payload,
    ...opts,
  };
}

/** Get the canonical JSON string for signing (deterministic key order) */
export function canonicalPayload(env: Envelope): string {
  const { id, type, from, to, timestamp, payload } = env;
  return JSON.stringify({ id, type, from, to, timestamp, payload });
}

/** Basic envelope validation */
export function validateEnvelope(env: unknown): env is Envelope {
  const e = env as Envelope;
  return (
    typeof e?.aip === 'string' &&
    typeof e?.id === 'string' &&
    typeof e?.type === 'string' &&
    typeof e?.from === 'string' &&
    typeof e?.to === 'string' &&
    typeof e?.timestamp === 'string' &&
    e?.payload !== undefined
  );
}
